Allow HPWH and Zonal ERVs to be modeled with AFNs

lymereJ opened the pull request:
Pull request overview
---------------------
 - Fixes #8503
 - This pull request excludes zonal ERV and HPWH nodes from the airflow network node verification so they can be modeled along with airflow networks.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

lymereJ committed:
Allow HPWH and Zonal ERV when modeling an AFN. Update dry air thermal conductivity correlation.

lymereJ committed:
Fix typo.

lymereJ committed:
Add logic to exclude standalone ERV with unbalanced airflow to be model along with AFN.

lymereJ commented on src/EnergyPlus/AirflowNetwork/src/Properties.cpp:
```diff
@@ -59,15 +59,15 @@ namespace AirflowNetwork {
 
-        Real64 const LowerLimit = -20;
-        Real64 const UpperLimit = 70;
+        Real64 const LowerLimit = -73;
+        Real64 const UpperLimit = 76.9;
 
-        Real64 const a = 0.02364;
-        Real64 const b = 0.0000754772569209165;
-        Real64 const c = -2.40977632412045e-8;
+        Real64 const a = 0.02420195316;
+        Real64 const b = 0.0000766872;
+        Real64 const c = -4.4e-08;
```
Update correlation reference, limits, and correlation coefficients.

lgu1234 replied:
You need to provide references to show how these coefficients are obtained. If from extrapolation, you need to provide justification.

lymereJ commented on src/EnergyPlus/AirflowNetworkBalanceManager.cc:
```diff
@@ -9584,6 +9588,12 @@ namespace AirflowNetworkBalanceManager {
             // Skip the inlet and outlet nodes of zone dehumidifiers
             if (GetZoneDehumidifierNodeNumber(state, i)) NodeFound(i) = true;
 
+            // Skip HPWH nodes that don't have to be included in the AFN
+            if (HeatPumpWaterHeaterNodeAFNExceptions(state, i)) NodeFound(i) = true;
+
+            // Skip Standalone ERV nodes that don't have to be included in the AFN
+            if (StandAloneERVAFNExceptions(state, i)) NodeFound(i) = true;
+
```
Add similar logic to `GetZoneDehumidifierNodeNumber` that provides an exception for reasonable standalone and HPWH configurations that can be modeled along an AFN without having to be explicitly part of the AFN.

lgu1234 replied:
Functionalities of GetZoneDehumidifierNodeNumber (exist), HeatPumpWaterHeaterNodeAFNExceptions (new), and StandAloneERVAFNExceptions (new) are the same. However, the function names are much different. I prefer to have similar function names, so that it will be easy for people to recognize the same functionality.

lymereJ commented on src/EnergyPlus/HVACStandAloneERV.cc:
```diff
@@ -2047,4 +2048,95 @@ namespace EnergyPlus::HVACStandAloneERV {
+                }
+            }
+
+            // If a standalone ERV's airflow is unbalanced it shouldn't be model along with an AFN
+            if (std::abs(SupplyFanAirFlow - ExhaustFanAirFlow) >= 1E-20 ||
+                std::abs(StandAloneERV.DesignSAFanVolFlowRate - StandAloneERV.DesignEAFanVolFlowRate) >= 1E-20) {
+                break;
+            }
```
Only allows standalone ERVs to be modeled if they have a balanced airflow.

lymereJ commented on src/EnergyPlus/WaterThermalTanks.cc:
```diff
@@ -11855,6 +11855,93 @@ namespace WaterThermalTanks {
+            // Get heat pump water heater data
+            HeatPumpWaterHeaterData &HPWH = state.dataWaterThermalTanks->HPWaterHeater(HeatPumpWaterHeaterIndex);
+
+            // "Zone and outdoor air" configuration is expected to change the zone pressure balance
+            if (HPWH.InletAirConfiguration != AmbientTempEnum::ZoneAndOA) {
```
The "Zone and outdoor air" configuration is expected to change the zone pressure balance so should probably not be modeled with an AFN.

lymereJ committed:
Supplement comment.

lgu1234 commented:
In general, one issue solves a single problem. It is better to separate your PR into 2 issues, because they are not related.

lymereJ committed:
Revert AFN air properties changes to move to a different PR.

mjwitte commented on src/EnergyPlus/HVACStandAloneERV.cc:
```diff
@@ -2047,4 +2048,96 @@ namespace EnergyPlus::HVACStandAloneERV {
 
+    bool StandAloneERVAFNExceptions(EnergyPlusData &state, int const NodeNumber)
+    {
+        // FUNCTION INFORMATION:
+        //       AUTHOR         Jeremy Lerond & Yan Chen
+        //       DATE WRITTEN   January 2021
+        //       MODIFIED       na
+        //       RE-ENGINEERED  na
+
```
Just a quick initial comment. We don't add author comments to new code anymore. Please delete these lines and similar ones in WaterThermalTanks. And this will require a unit test. I've marked this as a draft PR until that is completed. Then you can flip back to ready for review (that button shows at the bottom of the PR).

lymereJ replied:
Thanks @mjwitte for the review! Good to know, I just removed these lines, I was using `GetZoneDehumidifierNodeNumber` as a template for formatting (which includes them). We've been putting the tests on hold as Jason mentioned by email that the PR has issues. Should we go ahead and add them anyway? Or, should we keep holding off.

lymereJ committed:
Remove author names.

lymereJ committed:
Add new IDD entry, add documentation, add tests, rename functions, edit comments, moved an idf statement, added test files.

lymereJ commented:
@mjwitte, @lgu1234, @jasondegraw, @yanchenpnnl - I've updated the PR with the following changes:
- Added a new IDD field,
- Added documentation for the new field,
- Added tests (6 of them, all passed locally),
- Renamed the new functions,
- Minor edits to some comments
- Minor code modifications (moved an `if` statement)
- Added test IDF files

I *think* that it is ready for your review. It looks like I looks like I do not have the permissions to add the `IDDchange` label.

lymereJ committed:
ASSER_TRHOW to catch runtime error was not pushed.

lgu1234 commented on doc/input-output-reference/src/overview/group-airflow-network.tex:
```diff
@@ -326,6 +326,10 @@ \subsubsection{Inputs}\label{inputs-004}
+\paragraph{Field: Allow Unsupported Zone Equipment}\label{allow-unsupported-zone-equipment}
+
+This is an optional field. Input is Yes or No. The default is No. Set this input to Yes to have zone equipment that are currently unsupported in the AirflowNetwork model allowed in the simulation. Setting this field to Yes, allows the following equipment to be modeled along an AirflowNetwork model: ZoneHVAC:Dehumidifier, ZoneHVAC:EnergyRecoveryVentilator, WaterHeater:HeatPump:*.
+
```
I prefer to add a note:

The AFN model will exclude mass balance in these equipment objects and assume the mass flows are self-balanced in the equipment objects.

lymereJ replied:
@lgu1234, thanks! I've added the note.

lymereJ committed:
Expand new doc entry.

jasondegraw reviewed (approved):
I don't have time to pull this and build it, so someone should still do that, but based upon the discussions and calls, it looks to me like the main concern for me has been addressed (keeping the current behavior as the default, allow unsupported equipment with a flag). I'm not sure about the name of the variable, but I do like the name of the input field.

I might have time tomorrow or Friday to pull it, build it, and kick the tires, but I also might not.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -23835,12 +23835,21 @@ AirflowNetwork:SimulationControl,
+      \type choice
+      \key Yes
+      \key No
+      \default No
```
New field added to the end, :+1:, no transition required.

Myoldmopar commented on src/EnergyPlus/AirflowNetworkBalanceManager.cc:
```diff
@@ -9694,6 +9721,12 @@ namespace AirflowNetworkBalanceManager {
+        }
+        if (StandaloneERVFound) {
+            ShowWarningError(state, "ValidateDistributionSystem: A ZoneHVAC:EnergyRecoveryVentilator is simulated along with an AirflowNetwork but is not included in the AirflowNetwork.");
+        }
```
OK

Myoldmopar commented on src/EnergyPlus/HVACStandAloneERV.cc:
```diff
@@ -2047,4 +2048,91 @@ namespace EnergyPlus::HVACStandAloneERV {
+                // Fan:SystemModel
+                SupplyFanInletNodeIndex = HVACFan::fanObjs[StandAloneERV.SupplyAirFanIndex]->inletNodeNum;
+                SupplyFanOutletNodeIndex = HVACFan::fanObjs[StandAloneERV.SupplyAirFanIndex]->outletNodeNum;
+            	SupplyFanAirFlow = HVACFan::fanObjs[StandAloneERV.SupplyAirFanIndex]->designAirVolFlowRate;
```
Indentation problem here.

Myoldmopar commented on src/EnergyPlus/HVACStandAloneERV.cc:
```diff
@@ -2047,4 +2048,91 @@ namespace EnergyPlus::HVACStandAloneERV {
+        // Return value
+        bool StandSloneERVAFNException;
+
+        int StandAloneERVIndex;
```
On brand new code, there's no reason to follow these fortran patterns of declaring the variables up front.  Not going to hold this up for now, but just noting for the future.

Myoldmopar commented on testfiles/SingleFamilyHouse_HP_Slab.idf:
```diff
@@ -0,0 +1,5453 @@
+!-Generator IDFEditor 1.51
```
This file hasn't been added to the `testfiles/CMakeLists.txt` file, so this has never been run by our automated tests.  It will be interesting to see if it has errors/warnings once they are added.

OK, I went ahead and added them to the CMakeLists file, and ran them.  The error files have lots of messages, including a severe error, that need to be cleaned up before 9.5 is released.  @lymereJ please open an issue to track this and any other issues that need to be addressed before 9.5.

Myoldmopar reviewed (commented):
OK so a few things.

This pull request was created from the develop branch on LymereJ.  I am _very_ hesitant to push my updates to the develop branch of another repository, so that's inconvenient.  In the future, only open PRs from non-trunk branches.  

About the changes.  The code changes here look good and I think address the needs of the request.  But there were some changes I had to make to get it to pass our tests once they get into our main branch.  Firstly, the IDFs that were added were not listed in the `CMakeLists.txt` file, so they were not exercised.  Once added, I noticed they had a number of errors that needed to be cleaned out.  That cleaning process will need to happen after IO freeze is done, but before 9.5 is released.  

In order to handle the branching situation, I am going to push a new branch to the NREL/EnergyPlus repo, open up a brief PR that references back to this one for historical purposes, and merge it in with the changes including my updates.  Then I will close this PR with unmerged changes.

lymereJ commented:
@Myoldmopar - Thanks for the feedback and merging the changes, really appreciated!

> This pull request was created from the develop branch on LymereJ. I am _very_ hesitant to push my updates to the develop branch of another repository, so that's inconvenient. In the future, only open PRs from non-trunk branches.

I totally understand. I noticed that after creating the PR, I had made the same mistake in #7426. I'll keep that in mind for the next one (if there will be one).

> About the changes. The code changes here look good and I think address the needs of the request. But there were some changes I had to make to get it to pass our tests once they get into our main branch. Firstly, the IDFs that were added were not listed in the `CMakeLists.txt` file, so they were not exercised. Once added, I noticed they had a number of errors that needed to be cleaned out. That cleaning process will need to happen after IO freeze is done, but before 9.5 is released.

Noted and issue added: #8551.

> In order to handle the branching situation, I am going to push a new branch to the NREL/EnergyPlus repo, open up a brief PR that references back to this one for historical purposes, and merge it in with the changes including my updates. Then I will close this PR with unmerged changes.

Sounds good!
